File: packages/sdk-core/src/transport.ts
import { gzip } from 'pako';
import type { SpanData } from './types.js';

const RETRYABLE_STATUS = new Set([429, 500, 502, 503, 504]);
const MAX_RETRIES = 3;
const BASE_DELAY_MS = 250;

/**
 * Sends a batch of spans to the AgentLens ingest endpoint over HTTPS.
 *
 * Features:
 * - Gzip-compresses the JSON payload with `pako` (works in Node and browsers).
 * - Retries up to 3 times with exponential backoff on 429 / 5xx responses.
 * - Throws on final failure so the caller (buffer) can log and swallow.
 */
export class Transport {
  private readonly endpoint: string;
  private readonly apiKey: string;

  constructor(endpoint: string, apiKey: string) {
    this.endpoint = endpoint.replace(/\/$/, '');
    this.apiKey = apiKey;
  }

  /**
   * Posts `spans` to `{endpoint}/v1/spans`.
   * Resolves when the server returns 2xx; throws otherwise (after retries).
   */
  async send(spans: SpanData[]): Promise<void> {
    const compressed = this.compress(spans);
    const body: ArrayBuffer = compressed.buffer.slice(
      compressed.byteOffset,
      compressed.byteOffset + compressed.byteLength,
    ) as ArrayBuffer;

    for (let attempt = 0; attempt <= MAX_RETRIES; attempt++) {
      let response: Response;
      try {
        response = await fetch(`${this.endpoint}/v1/spans`, {
          method: 'POST',
          headers: {
            'Content-Type': 'application/json',
            'Content-Encoding': 'gzip',
            'X-API-Key': this.apiKey,
          },
          body,
        });
      } catch (networkErr) {
        // Network error (DNS, connection refused, timeout) — always retry
        if (attempt === MAX_RETRIES) throw networkErr;
        await this.delay(attempt);
        continue;
      }

      if (response.ok) return;

      if (RETRYABLE_STATUS.has(response.status) && attempt < MAX_RETRIES) {
        const retryAfterMs = this.parseRetryAfter(response) ?? this.backoff(attempt);
        await this.delay(attempt, retryAfterMs);
        continue;
      }

      // Non-retryable error
      throw new Error(
        `AgentLens transport: HTTP ${response.status} ${response.statusText}`,
      );
    }
  }

  // ── Private helpers ─────────────────────────────────────────────────────────

  private compress(spans: SpanData[]): Uint8Array {
    const json = JSON.stringify({ spans });
    return gzip(json);
  }

  private backoff(attempt: number): number {
    // Exponential backoff with full jitter: [0, base * 2^attempt)
    const cap = BASE_DELAY_MS * 2 ** attempt;
    return Math.random() * cap;
  }

  private parseRetryAfter(res: Response): number | undefined {
    const header = res.headers.get('Retry-After');
    if (!header) return undefined;
    const seconds = parseFloat(header);
    return isNaN(seconds) ? undefined : seconds * 1_000;
  }

  private delay(attempt: number, ms?: number): Promise<void> {
    return new Promise((resolve) =>
      setTimeout(resolve, ms ?? this.backoff(attempt)),
    );
  }
}

File: packages/sdk-python/agentlens/transport.py
"""HTTP transport — sends span batches to the AgentLens ingest endpoint.

Features:
- gzip-compresses the JSON payload using :mod:`gzip`.
- Retries up to 3 times with exponential backoff + full jitter on 429 / 5xx.
- Exposes both a synchronous :meth:`Transport.send` (used by the background
  buffer thread) and an async :meth:`Transport.send_async` (used by
  ``AgentLens.flush()`` when called from async code).
"""

from __future__ import annotations

import asyncio
import gzip
import json
import logging
import random
import time
from typing import Any

import httpx

logger = logging.getLogger(__name__)

_RETRYABLE_STATUS = {429, 500, 502, 503, 504}
_MAX_RETRIES = 3
_BASE_DELAY_S = 0.25


class Transport:
    """Sends compressed span batches to ``{endpoint}/v1/spans``.

    Args:
        endpoint: Base URL of the AgentLens ingest endpoint.
        api_key: Raw API key sent as ``X-API-Key`` header.
    """

    def __init__(self, endpoint: str, api_key: str) -> None:
        self._endpoint = endpoint.rstrip("/")
        self._api_key = api_key

    # ── Sync (used by background buffer thread) ─────────────────────────────

    def send(self, spans: list[dict[str, Any]]) -> None:
        """Send *spans* synchronously.  Raises on final failure."""
        body = self._compress(spans)
        headers = self._headers()
        url = f"{self._endpoint}/v1/spans"

        last_exc: Exception = RuntimeError("No attempt made")
        for attempt in range(_MAX_RETRIES + 1):
            try:
                with httpx.Client(timeout=10.0) as client:
                    resp = client.post(url, content=body, headers=headers)
                if resp.is_success:
                    return
                if resp.status_code in _RETRYABLE_STATUS and attempt < _MAX_RETRIES:
                    delay = self._backoff(attempt, resp.headers.get("Retry-After"))
                    time.sleep(delay)
                    continue
                resp.raise_for_status()
            except httpx.TransportError as exc:
                last_exc = exc
                if attempt < _MAX_RETRIES:
                    time.sleep(self._backoff(attempt))
                    continue
        raise last_exc

    # ── Async (used by AgentLens.flush() in async contexts) ─────────────────

    async def send_async(self, spans: list[dict[str, Any]]) -> None:
        """Send *spans* asynchronously.  Raises on final failure."""
        body = self._compress(spans)
        headers = self._headers()
        url = f"{self._endpoint}/v1/spans"

        last_exc: Exception = RuntimeError("No attempt made")
        for attempt in range(_MAX_RETRIES + 1):
            try:
                async with httpx.AsyncClient(timeout=10.0) as client:
                    resp = await client.post(url, content=body, headers=headers)
                if resp.is_success:
                    return
                if resp.status_code in _RETRYABLE_STATUS and attempt < _MAX_RETRIES:
                    delay = self._backoff(attempt, resp.headers.get("Retry-After"))
                    await asyncio.sleep(delay)
                    continue
                resp.raise_for_status()
            except httpx.TransportError as exc:
                last_exc = exc
                if attempt < _MAX_RETRIES:
                    await asyncio.sleep(self._backoff(attempt))
                    continue
        raise last_exc

    # ── Private helpers ──────────────────────────────────────────────────────

    def _compress(self, spans: list[dict[str, Any]]) -> bytes:
        payload = json.dumps({"spans": spans}).encode()
        return gzip.compress(payload)

    def _headers(self) -> dict[str, str]:
        return {
            "Content-Type": "application/json",
            "Content-Encoding": "gzip",
            "X-API-Key": self._api_key,
        }

    def _backoff(self, attempt: int, retry_after: str | None = None) -> float:
        if retry_after is not None:
            try:
                return float(retry_after)
            except ValueError:
                pass
        cap = _BASE_DELAY_S * (2**attempt)
        return random.random() * cap  # full jitter
